media/js/base/datalayer-trackscroll.es6.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+if (typeof window.dataLayer === 'undefined') {
+    window.dataLayer = [];
+}
+
+const TrackScroll = {};
+
+// track when page has been scrolled these percentages:
+let thresholds = [25, 50, 75, 90];
+// variables to support both throttle and debounce
+const pushDelay = 200;
+let lastPush = 0;
+let pushTimer;
+let listening = true;
+
+// get what percentage of the page has been scrolled
+TrackScroll.getDepth = () => {
+    const scrollHeight = document.documentElement.scrollHeight;
+    const innerHeight = window.innerHeight;
+    const scrollY = window.scrollY;
+    const scrollPosition = innerHeight + scrollY;
+    const depth = (100 * scrollPosition) / scrollHeight;
+    return depth;
+};
+
+// log the event to the dataLayer
+TrackScroll.sendEvent = (threshold) => {
+    window.dataLayer.push({
+        event: 'scroll',
+        percent_scrolled: String(threshold)
+    });
+};
+
+// removes the event listener after we're done
+TrackScroll.removeListener = () => {
+    if (listening) {
+        window.removeEventListener('scroll', TrackScroll.scrollListener, false);
+        listening = false;
+    }
+};
+
+// delayed call to scrollHandler, will keep getting delayed if it keeps getting called (aka this is a debouced call)
+TrackScroll.delayedScrollHandler = () => {
+    clearTimeout(pushTimer);
+    pushTimer = setTimeout(function () {
+        TrackScroll.scrollHandler();
+    }, pushDelay);
+};
+
+TrackScroll.scrollHandler = () => {
+    // check the browser supports filter before doing anything else
+    if (typeof Array.prototype.filter === 'function') {
+        const currentDepth = TrackScroll.getDepth();
+
+        // get a list of thresholds we've passed
+        const matchingThresholds = thresholds.filter(
+            (threshold) => currentDepth >= threshold
+        );
+
+        // remove thresholds we've passed from list of ones we're looking for
+        thresholds = thresholds.filter((threshold) => currentDepth < threshold);
+
+        // send the event for thresholds we passed
+        matchingThresholds.forEach((threshold) => {
+            TrackScroll.sendEvent(threshold);
+        });
+
+        // remove the event listener if we've scrolled past all thresholds
+        if (thresholds.length === 0) {
+            TrackScroll.removeListener();
+        }
+    } else {
+        // if it's too old to support logging, remove the listener
+        TrackScroll.removeListener();
+    }
+};
+
+TrackScroll.scrollListener = () => {
+    const now = new Date();
+
+    // if it's been a while since the last one, log it
+    if (now - lastPush >= pushDelay) {
```

Comment:
**Question:** I'm a bit curious why we need to do this frequent date checking / parsing here? This kind of negates the benefit of trying to de-bounce events a bit since it still fires continuously. 

Could we try something like this perhaps?

```javascript
TrackScroll.scrollListener = () => {
    clearTimeout(pushTimer);
    pushTimer = setTimeout(function () {
        TrackScroll.scrollHandler();
    }, pushDelay);
};
```

This seems to work OK when I test locally (GA events are still sent OK after the scroll).

Reply:
I think it's important to log the events intermittently in case someone slowly scrolls down the page and then exits before the timeout finishes. The date check is a throttle instead of a debounce (though, there's also a debounce to catch the final state after scrolling stops). In theory the date check is less resource intensive than executing the full function continuously.

If you're concerned could we try the throttled function for a week and then move to one that is only debounced and compare the data. It's possible the throttle is unnecessary but I worry.

Reply:
> I think it's important to log the events intermittently in case someone slowly scrolls down the page and then exits before the timeout finishes.

I tried to to this using a pretty standard de-bounce with a 100ms delay, and i couldn't really get it not to send an event before i could close anything. Maybe if I used a keyboard shortcut? It does seem a little unlikely when trying it out practically.

>  In theory the date check is less resource intensive than executing the full function continuously.

I'm not 100% sure if we need this more complicated mix of throttling and de-bouncing though, when we're only sending a maximum of 4 events? 

It's been a while since I've had to think about this stuff, but the way de-bouncing usually works (iiuc) is by returning a closure, and then you can do either a `setTimeout` or a date comparison check (similar to what you have) - and all this should happen async, outside of the actual scroll event handler. That way there's no need to execute the full function continuously. 

If you're concerned about losing data, then lowering the delay to 100ms would still be a pretty big improvement imho (although 200ms seems to work ok for me still as well). Here's what I mean by returning a closure:

```javascript
function debounce(func, delay) {
    let timer;
    return function () {
        clearTimeout(timer);
        timer = setTimeout(() => {
            func.apply(this, arguments);
        }, delay);
    };
}

const onScroll = debounce(TrackScroll.scrollHandler, 100);

window.addEventListener('scroll', onScroll, false);
```